storage/zynq7000-flash/flashdrv.c

```diff
@@ -0,0 +1,502 @@
+	size_t cmdSz, dataSz, transferSz;
+	const flash_cmd_t *cmd = &fdrv_common.info.cmds[flash_cmd_qior];
+
+	if (len == 0)
+		return 0;
+
+	if (buff == NULL || (offs + len) > CFI_SIZE_FLASH(fdrv_common.info.cfi.chipSize))
+		return -EINVAL;
+
+	memset(fdrv_common.cmdTx, 0, cmd->size);
+	fdrv_common.cmdTx[0] = cmd->opCode;
+	fdrv_common.cmdTx[1] = (offs >> 16) & 0xff;
+	fdrv_common.cmdTx[2] = (offs >> 8) & 0xff;
+	fdrv_common.cmdTx[3] = offs & 0xff;
+
+	/* Cmd size is rounded to 4 bytes, it includes dummy cycles [b]. */
+	cmdSz = (cmd->size + cmd->dummyCyc / 8 + 0x3) & ~0x3;
+
+	/* Size of the data which is received during cmd transfer */
+	transferSz = cmdSz - cmd->size - cmd->dummyCyc / 8;
+	dataSz = (transferSz > len) ? len : transferSz;
+
+	qspi_start();
+	if ((res = qspi_transfer(fdrv_common.cmdTx, fdrv_common.cmdRx, cmdSz, TIMEOUT_CMD_MS)) < 0) {
+		qspi_stop();
+		return res;
+	}
+	/* Copy data received after dummy bytes */
+	memcpy(buff, (fdrv_common.cmdRx + cmd->size + cmd->dummyCyc / 8), dataSz);
+
+	res = qspi_transfer(NULL, (uint8_t *)buff + dataSz, len - dataSz, timeout);
+	qspi_stop();
+
+	return (res < 0) ? res : (res + dataSz);
+}
+
+
+ssize_t flashdrv_write(addr_t offs, const void *buff, size_t len)
+{
+	ssize_t res;
+	time_t timeout;
+	size_t chunkSz = 0, freeSz = 0, saveSz = 0;
+	static const uint32_t timeoutFactor = 0x100;
+
+	int regID, sectID;
+	size_t sectSz, regStart;
+
+	const flash_cfi_t *cfi = &fdrv_common.info.cfi;
+
+	if (len == 0)
+		return 0;
+
+	if (buff == NULL || (offs + len) > CFI_SIZE_FLASH(cfi->chipSize))
+		return -EINVAL;
+
+	while (saveSz < len) {
+		offs += chunkSz;
+
+		if ((regID = flashdrv_regionFind(offs)) < 0)
+			return -EINVAL;
+
+		regStart = flashdrv_regStart(regID);
+		sectSz = CFI_SIZE_SECTION(cfi->regs[regID].size);
+		sectID = (offs - regStart) / sectSz;
+
+		if (regID != fdrv_common.regID || sectID != fdrv_common.sectID) {
+			if ((res = flashdrv_sync()) < 0)
+				return res;
+
+			/* Read operation timeout depends on sector size. Factor value selected empirically. */
+			timeout = (sectSz * TIMEOUT_CMD_MS) / timeoutFactor;
+			if ((res = flashdrv_read(regStart + (sectSz * sectID), fdrv_common.buff, sectSz, timeout)) < 0)
+				return res;
+
+			if ((res = flashdrv_sectorErase(regStart + (sectSz * sectID))) < 0)
+				return res;
+
+			fdrv_common.regID = regID;
+			fdrv_common.sectID = sectID;
+			fdrv_common.buffPos = offs - regStart - (sectID * sectSz);
+		}
+
+		freeSz = sectSz - fdrv_common.buffPos;
+		chunkSz = (freeSz > (len - saveSz)) ? (len - saveSz) : freeSz;
+		memcpy(fdrv_common.buff + fdrv_common.buffPos, (const uint8_t *)buff + saveSz, chunkSz);
+
+		saveSz += chunkSz;
+		fdrv_common.buffPos += chunkSz;
+
+		if (fdrv_common.buffPos < sectSz)
+			continue;
+
+		if ((res = flashdrv_sync()) < 0)
+			return res;
+	}
+
+	return saveSz;
+}
+
+
+int flashdrv_done(void)
+{
+	int res = EOK;
+
+	do {
+		if ((res = flashdrv_sync()) < 0)
+			break;
+
+		if ((res = qspi_deinit()) < 0)
```

Comment:
We cannot do qpi_deinit() if flashdrv_sync() failed?

Reply:
Added `qpi_deinit() ` after failing `flashdrv_sync()`.